File: DktMalfLabeling/Inner/trainModel.R
library( ANTsR )
library( ANTsRNet )
library( keras )
library( tensorflow )
library( ggplot2 )

keras::backend()$clear_session()

Sys.setenv( "CUDA_VISIBLE_DEVICES" = "1" )

baseDirectory <- '/home/ntustison/Data/DktMalfLabeling/'
scriptsDirectory <- paste0( baseDirectory, 'Scripts/' )
source( paste0( scriptsDirectory, 'batchGenerator.R' ) )

templateDirectory <- '/home/ntustison/Data/BrainAge2/Data/Templates/'
template <- antsImageRead( paste0( templateDirectory, "croppedMNI152.nii.gz" ) )
templateSize <- c( 160L, 192L, 160L )

################################################
#
#  Create the model and load weights
#
################################################

#classes <- c( 0, 4, 6, 7, 10:18, 24, 26, 28, 30, 43, 44:46, 49:54, 58, 60, 91:92, 
#              630:632, 1002:1003, 1005:1031, 1034:1035, 2002:2003, 2005:2031, 2034:2035 )

classes <- c( 0, 4, 6, 7, 10:18, 24, 26, 28, 30, 43, 44:46, 49:54, 58, 60, 91:92,
              630:632 )

numberOfClassificationLabels <- length( classes )
imageModalities <- c( "T1" )
channelSize <- length( imageModalities )

unetModel <- createUnetModel3D( c( templateSize, channelSize ),
  numberOfOutputs = numberOfClassificationLabels, mode = 'classification',
  numberOfLayers = 4, numberOfFiltersAtBaseLayer = 8, dropoutRate = 0.0,
  convolutionKernelSize = c( 3, 3, 3 ), deconvolutionKernelSize = c( 2, 2, 2 ),
  weightDecay = 0.0, addAttentionGating = TRUE )

brainWeightsFileName <- paste0( scriptsDirectory, "/dktLabeling.h5" )
if( file.exists( brainWeightsFileName ) )
  {
  load_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )
  # } else {
  # stop( "Weights file doesn't exist.\n" )  
  }

metric_multilabel_dice_coefficient <-
  custom_metric( "multilabel_dice_coefficient",
    multilabel_dice_coefficient )

loss_dice <- function( y_true, y_pred ) {
   -multilabel_dice_coefficient(y_true, y_pred)
   }
attr(loss_dice, "py_function_name") <- "multilabel_dice_coefficient"

unetModel %>% compile(
  optimizer = optimizer_adam(),
  loss = loss_dice, # weighted_loss, # tensorflow::tf$keras$losses$CategoricalCrossentropy(),
  metrics = c( metric_multilabel_dice_coefficient, metric_categorical_crossentropy ) )

# unetModel %>% compile(
#  optimizer = optimizer_adam(),
#  loss = categorical_focal_loss( alpha = 0.25, gamma = 2.0 ),
#  metrics = 'accuracy' )

################################################
#
#  Load the brain data
#
################################################

cat( "Loading brain data.\n" )

brainImages <- c( 
  Sys.glob( "/raid/data_NT/CorticalThicknessData2014/*/ThicknessAnts/*xMniInverseWarped.nii.gz" ),
  Sys.glob( "/home/ntustison/Data/SixTissueSegmentation/Data/*/*xMniInverseWarped.nii.gz" ) )

trainingImageFiles <- c()
trainingSegmentationImageFiles <- c()

pb <- txtProgressBar( min = 0, max = length( brainImages ), style = 3 )
for( i in seq_len( length( brainImages ) ) )
  {
  setTxtProgressBar( pb, i )

  image <- brainImages[i]
  seg <- gsub( "InverseWarped", "RefinedMalfLabels", image )

  if( ! file.exists( image ) || ! file.exists( seg ) )
    {
    next  
    # stop( "Mask doesn't exist." )  
    }

  trainingImageFiles <- append( trainingImageFiles, image )
  trainingSegmentationImageFiles <- append( trainingSegmentationImageFiles, seg )
  }
cat( "\n" )  

cat( "Total training image files: ", length( trainingImageFiles ), "\n" )

cat( "\nTraining\n\n" )

###
#
# Set up the training generator
#

batchSize <- 1L

# Split trainingData into "training" and "validation" componets for
# training the model.

numberOfData <- length( trainingImageFiles )
sampleIndices <- sample( numberOfData )

validationSplit <- floor( 0.8 * numberOfData )
trainingIndices <- sampleIndices[1:validationSplit]
numberOfTrainingData <- length( trainingIndices )
validationIndices <- sampleIndices[( validationSplit + 1 ):numberOfData]
numberOfValidationData <- length( validationIndices )

###
#
# Run training
#

track <- unetModel %>% fit_generator(
  generator = batchGenerator( batchSize = batchSize,
                              imageSize = templateSize, 
                              template = template,
                              images = trainingImageFiles[trainingIndices],
                              segmentationImages = trainingSegmentationImageFiles[trainingIndices],
                              segmentationLabels = classes,
                              doRandomContralateralFlips = FALSE,
                              doDataAugmentation = TRUE 
                            ),
  steps_per_epoch = 100L,
  epochs = 200L,
  validation_data = batchGenerator( batchSize = batchSize,
                                    imageSize = templateSize, 
                                    template = template,
                                    images = trainingImageFiles[validationIndices],
                                    segmentationImages = trainingSegmentationImageFiles[validationIndices],
                                    segmentationLabels = classes,
                                    doRandomContralateralFlips = FALSE,
                                    doDataAugmentation = TRUE
                                  ),  
  validation_steps = 40L,
  callbacks = list(
    callback_model_checkpoint( brainWeightsFileName,
      monitor = 'val_loss', save_best_only = TRUE, save_weights_only = TRUE,
      verbose = 1, mode = 'auto' ), 
     callback_reduce_lr_on_plateau( monitor = 'val_loss', factor = 0.3,
       verbose = 1, patience = 10, mode = 'auto' ),
    callback_early_stopping( monitor = 'val_loss', min_delta = 0.001,
      patience = 20 )
  )
)

save_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )

File: XRay/LungExtraction/train_model.py
import ants
import antspynet

import os
os.environ["CUDA_VISIBLE_DEVICES"] = "3"
import glob

import random

import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K

from batch_generator import batch_generator

K.clear_session()
gpus = tf.config.experimental.list_physical_devices("GPU")
if len(gpus) > 0:
    tf.config.experimental.set_memory_growth(gpus[0], True)

tf.compat.v1.disable_eager_execution()

base_directory = '/home/ntustison/Data/XRayLungExtraction/'
scripts_directory = base_directory + 'Scripts/'


################################################
#
#  Create the model and load weights
#
################################################

number_of_labels = 3
image_modalities = ["XRay"]

image_size = (256, 256)
channel_size = 3

unet_model = antspynet.create_unet_model_2d((*image_size, channel_size),
   number_of_outputs=number_of_labels, mode="classification", 
   number_of_filters_at_base_layer=32, number_of_layers=4,
   convolution_kernel_size=(3, 3), deconvolution_kernel_size=(2, 2),
   dropout_rate=0.0, weight_decay=0,
   additional_options=None)

weighted_loss = antspynet.weighted_categorical_crossentropy(weights=(1, 2, 2))
dice_loss = antspynet.multilabel_dice_coefficient(dimensionality=2, smoothing_factor=0.)

weights_filename = scripts_directory + "xrayLungExtraction.h5"
if os.path.exists(weights_filename):
    unet_model.load_weights(weights_filename)

unet_model.compile(optimizer=tf.keras.optimizers.legacy.Adam(learning_rate=2e-4),
                    loss=dice_loss,
                    metrics=[dice_loss])

################################################
#
#  Load the data
#
################################################

print("Loading lung data.")

images = (*glob.glob(base_directory + "/**/CXR/*.nii.gz", recursive=True),
          *glob.glob(base_directory + "/**/CXR2/*.nii.gz", recursive=True),
          *glob.glob(base_directory + "/**/CXR3/*.nii.gz", recursive=True))

left_prior = ants.image_read(base_directory + "leftPrior.nii.gz", dimension=2)
right_prior = ants.image_read(base_directory + "rightPrior.nii.gz", dimension=2)

training_image_files = list()
training_label_files = list()

for i in range(len(images)):
    base_file = os.path.basename(images[i])
    local_dir = os.path.dirname(images[i])
    label_image = local_dir.replace("CXR", "Masks") + "/" + base_file
    print(images[i])
    print(label_image)
    if os.path.exists(label_image):
        training_image_files.append(images[i])
        training_label_files.append(label_image)

print("Total training image files: ", len(training_image_files))

print( "Training")


###
#
# Set up the training generator
#

batch_size = 32

generator = batch_generator(batch_size=batch_size,
                            images=training_image_files,
                            label_images=training_label_files,
                            left_prior=left_prior,
                            right_prior=right_prior,
                            do_histogram_intensity_warping=True,
                            do_simulate_bias_field=True,
                            do_add_noise=True,
                            do_data_augmentation=True
                            )

track = unet_model.fit(x=generator, epochs=2000, verbose=1, steps_per_epoch=32,
    callbacks=[
       keras.callbacks.ModelCheckpoint(weights_filename, monitor='loss',
           save_best_only=True, save_weights_only=True, mode='auto', verbose=1),
       keras.callbacks.ReduceLROnPlateau(monitor='loss', factor=0.95,
          verbose=1, patience=10, mode='auto'),
#       keras.callbacks.EarlyStopping(monitor='loss', min_delta=0.001,
#          patience=20)
       ]
   )

unet_model.save_weights(weights_filename)


